[VectorTypeInfo] Change default handling of data buffer

Camille-K opened the pull request:
In the `getValuePtr` methods, a buffer on vector data is obtained by actually returning a reference to the vector first element. However the case where the method is called on an empty (but valid) vector is not handled. This may lead to an execution error, for instance when using SofaPython3 to assign values to a component's Data dynamically. This specific scenario is the object of a second PR in SofaPython3 [(link to the PR)](https://github.com/sofa-framework/SofaPython3/pull/318).

To change this, we propose to use the `std::vector.data()` method, which handles the case when the vector is actually empty (however returning a pointer that may or may not be null).

All unit tests work on my setup with this modification. However I don't fully understand the consequences of a change at this depth in SOFA's core, so additional opinions would be useful.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

Commit:
[VectorTypeInfo] Change default handling of data buffer

Default behaviour of the getValuePtr methods was to give an address to
the vector data by returning a reference to the vector first element.
However the case where the method is called on a valid empty vector is
not handled.
To change this, we use the std::vector.data() method, which handles the
case when the vector is actually empty (returning a pointer that may or
may not be null)

fredroy commented:
[ci-build][with-all-tests]

alxbilger commented:
The problem with this function is that you cannot trust the returned value forever due to pointer invalidation.

`std::vector` allocates dynamically the memory depending of the need. At creation, it allocates a small chunk of memory. If the memory is full when you try to insert an element, it will allocate a bigger chunk of memory and move everything in this new memory space. Everything that was in the previous memory space is invalidated. It stills points to something, but it is invalid. And using it leads to a crash.

I made a small example on godbolt of this situation: https://godbolt.org/z/aqec8Krzo

You can see that the result of `data()` is invalidated after elements insertions.

Conclusion: your solution is better than the previous. But this function must be used appropriately.

Camille-K commented:
Thanks @alxbilger for the explanation and the perfect illustration :+1:
I understand the issue, although I don't know if the "appropriate use" of this method (`getValuePtr`) should be left to the user, or if additional safeguards should be put in the code.

I did a quick search in the code of SOFA + a few plugins that I'm using, and I found a call to this particular method only in the `DataHelper` component in SofaPython3. I added a quick safeguard in this case, in the [PR](https://github.com/sofa-framework/SofaPython3/pull/318) I'm referring to in the description above. However the modification I suggested doesn't solve the type of problem you are describing here (so it would be left to the user to check data buffer coherence whenever the underlying data is modified...).

Apart from that, I saw that the initial code I replaced (`return &data[0]`) is also used in two others `getValuePtr` methods : in `FixedArrayTypeInfo.h` and `TypeInfo_Text.h`. As I didn't find any other call to these methods, I would say this doesn't change much, but maybe the PR could be extended to these as well. I can do the modification if someone can confirm it is a good idea.

alxbilger commented:
It's definitively a good idea to also replace the `&data[0]` calls by `.data()`.

Camille-K commented:
I did the modification in `TypeInfo_Text.h`, but build fails if I extend it to `FixedArrayTypeInfo.h`, apparently because of the template specialisations on Rigid types, for instance :
```
FixedArrayTypeInfo.h:195: error: ‘const DataType’ {aka ‘const class sofa::defaulttype::RigidDeriv<2, double>’} has no member named ‘data’
  195 |         return data.data();
      |                ~~~~~^~~~
```
(which makes sense).

Commit:
[DataBufferPointer] Extend modification to TypeInfo_Text
